pyacq/core/stream.py

```diff
@@ -523,11 +600,386 @@ def get_array_slice(self, index, length, autoswapaxes = True):
+        
+        # by default, index advances by the size of the chunk
+        if index is None:
+            index = self._write_index + dsize
+        
+        assert dsize <= index - self._write_index, ("Data size is %d, but index "
+                                                    "only advanced by %d." % 
+                                                    (dsize, index-self._write_index)) 
+
+        revert_inds = [self._read_index, self._write_index]
+        try:
+            # advance write index. This immediately prevents other processes from
+            # accessing memory that is about to be overwritten.
+            self._set_write_index(index)
+            
+            # decide if any skipped data needs to be filled in
+            fill_start = max(self._read_index + 1, self._write_index + 1 - bsize)
+            fill_stop = self._write_index + 1 - dsize
+            
+            if fill_stop > fill_start:
+                # data was skipped; fill in missing regions with 0.
+                self._fill(fill_start, fill_stop)
+                revert_inds[1] = fill_stop-1
+                
+            self._write(self._write_index + 1 - dsize, self._write_index + 1, data)
+                
+            self._set_read_index(index)
+        except:
+            # If there is a failure writing data, revert read/write pointers
+            self._set_read_index(revert_inds[0])
+            self._set_write_index(revert_inds[1])
+            raise
+
+    def _write(self, start, stop, value):
+        # get starting index
+        bsize = self.nbytes
+        dsize = stop - start
+        i = start % bsize
+        
+        if self.double:
+            self.buffer[i:i+dsize] = value
+            i += bsize
+        
+        if i + dsize <= self.buffer.shape[0]:
+            self.buffer[i:i+dsize] = value
+        else:
+            n = self.buffer.shape[0]-i
+            self.buffer[i:] = value[:n]
+            self.buffer[:dsize-n] = value[n:]
+            
+    def _fill(self, start, stop):
+        self._write(start, stop, 0)
+
+
+class ArrayRingBuffer(RingBuffer):
+    """Array-like object that wraps an internal ring buffer.
+    
+    This allows the user to request the concatenated history of data
+    received by the stream, up to a predefined length. Double ring buffers
+    allow faster, copyless reads at the expense of doubled write time and memory
+    footprint.    
+    """
+    def __init__(self, shape, dtype, double=True, shmem=None, fill=None):
+        self.double = double
+        self.shape = shape
+        
+        shape = (shape[0] * (2 if double else 1),) + shape[1:]
+        # initialize int buffers with 0 and float buffers with nan
+        if fill is None:
+            fill = 0 if np.dtype(dtype).kind in 'ui' else np.nan
+        self._filler = fill
+        
+        bsize = np.product(self.shape) * np.dtype(dtype).itemsize
+        RingBuffer.__init__(self, bsize, double=double, shmem=shmem)
```

Comment:
No, you're right--ArrayRingBuffer is forced to have timeaxis=0 internally. But then, it does not make any sense to have an array ring buffer that works any other way. If you want to be able to stack array chunks in a ring buffer and be able to extract any segment of the buffer without copy, then you _must_ have timeaxis=0 in the ring buffer.

So my plan is something like:

* Streams send / recv array chunks without altering their internal memory layout--if channels are not interleaved when they are sent, then they will not be interleaved when they are received, regardless of the stream type.
* Sharedmem streams use RingBuffer to store byte-arrays in any axis order
* If you attach an ArrayRingBuffer to an InputStream, then by default it will use its own memory buffer and copy chunks into the buffer with timeaxis=0 as they arrive.
* If a sharedmem stream's buffer already happens to have its memory layout with timeaxis=0, then the ArrayRingBuffer can use the mmap buffer rather than creating its own memory and copying.